File: java-db-sample/src/main/java/com/example/java/repository/UserRepository.java
package com.example.java.repository;

import com.example.java.entity.User;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Java版のRepositoryクラス
 * DBアクセスロジックを実装
 */
public class UserRepository {
    private final Connection connection;

    public UserRepository(Connection connection) {
        this.connection = connection;
    }

    public void createTable() throws SQLException {
        String sql = "CREATE TABLE IF NOT EXISTS users (" +
                "id BIGINT AUTO_INCREMENT PRIMARY KEY, " +
                "name VARCHAR(255) NOT NULL, " +
                "email VARCHAR(255) NOT NULL, " +
                "age INT NOT NULL, " +
                "nickname VARCHAR(255), " +
                "phone_number VARCHAR(255), " +
                "address VARCHAR(255)" +
                ")";
        try (Statement stmt = connection.createStatement()) {
            stmt.execute(sql);
        }
    }

    public User save(User user) throws SQLException {
        String sql = "INSERT INTO users (name, email, age, nickname, phone_number, address) VALUES (?, ?, ?, ?, ?, ?)";
        try (PreparedStatement pstmt = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            pstmt.setString(1, user.getName());
            pstmt.setString(2, user.getEmail());
            pstmt.setInt(3, user.getAge());
            pstmt.setString(4, user.getNickname());
            pstmt.setString(5, user.getPhoneNumber());
            pstmt.setString(6, user.getAddress());
            pstmt.executeUpdate();

            try (ResultSet generatedKeys = pstmt.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    user.setId(generatedKeys.getLong(1));
                }
            }
        }
        return user;
    }

    public Optional<User> findById(Long id) throws SQLException {
        String sql = "SELECT id, name, email, age, nickname, phone_number, address FROM users WHERE id = ?";
        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            pstmt.setLong(1, id);
            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    return Optional.of(mapResultSetToUser(rs));
                }
            }
        }
        return Optional.empty();
    }

    public List<User> findAll() throws SQLException {
        List<User> users = new ArrayList<>();
        String sql = "SELECT id, name, email, age, nickname, phone_number, address FROM users";
        try (Statement stmt = connection.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {
            while (rs.next()) {
                users.add(mapResultSetToUser(rs));
            }
        }
        return users;
    }

    public void update(User user) throws SQLException {
        String sql = "UPDATE users SET name = ?, email = ?, age = ?, nickname = ?, phone_number = ?, address = ? WHERE id = ?";
        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            pstmt.setString(1, user.getName());
            pstmt.setString(2, user.getEmail());
            pstmt.setInt(3, user.getAge());
            pstmt.setString(4, user.getNickname());
            pstmt.setString(5, user.getPhoneNumber());
            pstmt.setString(6, user.getAddress());
            pstmt.setLong(7, user.getId());
            pstmt.executeUpdate();
        }
    }

    public void deleteById(Long id) throws SQLException {
        String sql = "DELETE FROM users WHERE id = ?";
        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            pstmt.setLong(1, id);
            pstmt.executeUpdate();
        }
    }

    /**
     * ResultSetからUserオブジェクトへの変換
     *
     * Null安全性の課題:
     * - getString()はnullを返す可能性があるが、Javaでは型システムで保証されない
     * - nullチェックを忘れるとNullPointerExceptionのリスク
     * - Kotlinと異なり、コンパイル時にnullチェックが強制されない
     */
    private User mapResultSetToUser(ResultSet rs) throws SQLException {
        return new User(
                rs.getLong("id"),
                rs.getString("name"),
                rs.getString("email"),
                rs.getInt("age"),
                rs.getString("nickname"),       // nullの可能性があるが、型システムでは検出されない
                rs.getString("phone_number"),   // nullの可能性があるが、型システムでは検出されない
                rs.getString("address")         // nullの可能性があるが、型システムでは検出されない
        );
    }
}

File: kotlin-db-sample/src/main/kotlin/com/example/kotlin/repository/UserRepository.kt
package com.example.kotlin.repository

import com.example.kotlin.entity.User
import java.sql.Connection
import java.sql.ResultSet
import java.sql.Statement

/**
 * Kotlin版のRepositoryクラス
 * DBアクセスロジックを実装
 *
 * Kotlinの特徴:
 * - use関数でリソースの自動クローズ
 * - 拡張関数でResultSetからUserへの変換を簡潔に
 * - letやrunなどのスコープ関数で処理をチェーン
 * - nullableな型とエルビス演算子でNull安全
 */
class UserRepository(private val connection: Connection) {

    fun createTable() {
        val sql = """
            CREATE TABLE IF NOT EXISTS users (
                id BIGINT AUTO_INCREMENT PRIMARY KEY,
                name VARCHAR(255) NOT NULL,
                email VARCHAR(255) NOT NULL,
                age INT NOT NULL,
                nickname VARCHAR(255),
                phone_number VARCHAR(255),
                address VARCHAR(255)
            )
        """.trimIndent()
        connection.createStatement().use { it.execute(sql) }
    }

    fun save(user: User): User {
        val sql = "INSERT INTO users (name, email, age, nickname, phone_number, address) VALUES (?, ?, ?, ?, ?, ?)"
        connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS).use { pstmt ->
            pstmt.setString(1, user.name)
            pstmt.setString(2, user.email)
            pstmt.setInt(3, user.age)
            pstmt.setString(4, user.nickname)
            pstmt.setString(5, user.phoneNumber)
            pstmt.setString(6, user.address)
            pstmt.executeUpdate()

            pstmt.generatedKeys.use { rs ->
                if (rs.next()) {
                    user.id = rs.getLong(1)
                }
            }
        }
        return user
    }

    fun findById(id: Long): User? {
        val sql = "SELECT id, name, email, age, nickname, phone_number, address FROM users WHERE id = ?"
        return connection.prepareStatement(sql).use { pstmt ->
            pstmt.setLong(1, id)
            pstmt.executeQuery().use { rs ->
                if (rs.next()) rs.toUser() else null
            }
        }
    }

    fun findAll(): List<User> {
        val sql = "SELECT id, name, email, age, nickname, phone_number, address FROM users"
        return connection.createStatement().use { stmt ->
            stmt.executeQuery(sql).use { rs ->
                buildList {
                    while (rs.next()) {
                        add(rs.toUser())
                    }
                }
            }
        }
    }

    fun update(user: User) {
        val sql = "UPDATE users SET name = ?, email = ?, age = ?, nickname = ?, phone_number = ?, address = ? WHERE id = ?"
        connection.prepareStatement(sql).use { pstmt ->
            pstmt.setString(1, user.name)
            pstmt.setString(2, user.email)
            pstmt.setInt(3, user.age)
            pstmt.setString(4, user.nickname)
            pstmt.setString(5, user.phoneNumber)
            pstmt.setString(6, user.address)
            pstmt.setLong(7, user.id ?: throw IllegalArgumentException("User ID must not be null"))
            pstmt.executeUpdate()
        }
    }

    fun deleteById(id: Long) {
        val sql = "DELETE FROM users WHERE id = ?"
        connection.prepareStatement(sql).use { pstmt ->
            pstmt.setLong(1, id)
            pstmt.executeUpdate()
        }
    }

    /**
     * 拡張関数: ResultSetからUserオブジェクトへの変換
     * Java版より簡潔に書ける
     *
     * Null安全性:
     * - getString()はnullを返す可能性があるため、String?型として扱う
     * - KotlinのgetString()はnullableなのでそのままString?に代入可能
     */
    private fun ResultSet.toUser() = User(
        id = getLong("id"),
        name = getString("name"),
        email = getString("email"),
        age = getInt("age"),
        nickname = getString("nickname"),          // DBからnullが返される可能性
        phoneNumber = getString("phone_number"),   // DBからnullが返される可能性
        address = getString("address")             // DBからnullが返される可能性
    )
}
